### PR TITLE
Apriltag vision

### LINKED ISSUE #3: Full field vision and odometry backfeeding
As the title suggests, using the new apriltag vision system to create a full vision pipeline

- [x] use this data (IDs, location) in order to determine robot odometry absolutely based purely on odometry
- [x] factor in this vision into robot odometry in a weighted fasion
- [x] simulate field vision by using the known locations in a simulated fasion to mimic how the vision system outputs data
- [x] use photonvision to get positions for a real robot
- [x] update to 2023 positioning

### LINKED ISSUE #16: Reorient gyro based on apriltag positioning
the apriltags on the field are in fixed positions, so preventing drifting can be done via the relative positions

### PR DESCRIPTION
**Pull request by lmaxwell24**
construct the vision system in preperation for the 2023 vision pipeline

this does NOT implement photonvision direct support, however contains simulation and publishing

### CODE REVIEW

**Comment by lmaxwell24** [comment]
now contains photonvision direct support using photonlib

File: constants.py [hunk 1]
```diff
@@ -201,6 +208,10 @@
 kLimelightTargetVerticalAngleKey = "ty"
 kLimelightLEDModeKey = "ledMode"
 kLimelightTrackerModuleName = "limelight"
+kLimelightRelativeToRobotTransform = Transform3d(
+    Pose3d(),
+    Pose3d(0.258, 0, 1.01, Rotation3d()),
```
Comment (LandonBayer) [inline constants.py:213]: Where do these numbers come from?

File: physics.py [hunk 2]
```diff
@@ -9,7 +9,10 @@
 # of your robot code without too much extra effort.
 #
 
+import functools
+import operator
```
Comment (LandonBayer) [inline physics.py:13]: Are these in robotpy? I assume so but want to ask to make sure it'll work on stuff

File: subsystems/drivesubsystem.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
     SwerveDrive4Odometry,
     SwerveModulePosition,
 )
+from wpimath.estimator import SwerveDrive4PoseEstimator
```
Comment (LandonBayer) [inline subsystems/drivesubsystem.py:30]: Is this new? There was a bunch of math and stuff last time

File: subsystems/visionsubsystem.py [hunk 4]
```diff
@@ -1,9 +1,90 @@
...
+        self.camera = PhotonCamera(
+            NetworkTableInstance.getDefault(), constants.kPhotonvisionCameraName
+        )
+
+    def getCameraToTargetTransforms(self) -> List[Tuple[int, Transform3d]]:
+        """this function returns a list of the type (target_id, transformCameraToTarget) for every target"""
+        photonResult = self.camera.getLatestResult()
+        if photonResult.hasTargets():
+            return [
+                (target.getFiducialId(), target.getBestCameraToTarget())
```
Comment (LandonBayer) [inline subsystems/visionsubsystem.py:32]: Where is the variable "target" coming from, is it from the function?

**Review by LandonBayer — DISMISSED** [review]

File: constants.py (hunk 1 above)
Comment (lmaxwell24) [reply constants.py:213]: cad

File: physics.py (hunk 2 above)
Comment (lmaxwell24) [reply physics.py:13]: default python

File: subsystems/drivesubsystem.py (hunk 3 above)
Comment (lmaxwell24) [reply subsystems/drivesubsystem.py:30]: yep the pose estimator is new from wpilib

File: constants.py [hunk 5]
```diff
@@ -337,7 +348,7 @@
 
 kRobotPoseArrayKeys = OptionalValueKeys("RobotPoseArray")
 
-kRobotVisionPoseWeight = 0.19  # 19% vision data
+kRobotVisionPoseWeight = 0.05  # 19% vision data
```
Comment (agelsomini) [inline constants.py]: I'm guessing the comment should be updated?

File: constants.py [hunk 6]
```diff
@@ -357,6 +368,56 @@
 
 kTargetName = "Target"
 
+kApriltagPositionDict = {  # this is gathered from a mapping of the field, derived with positions and ZYX order rotations converted to Quaternions in bulk in a spreadsheet
```
Comment (agelsomini) [inline constants.py]: Is that spreadsheet in google drive?  Just want to make sure someone else can recreate it.

File: requirements.txt [hunk 7]
```diff
@@ -18,4 +18,4 @@ robotpy-navx==2023.0.1
 robotpy-wpimath==2023.1.1.0
 robotpy-wpiutil==2023.1.1.0
 wpilib==2023.1.1.0
-
+robotpy-photonvision==2023.1.1b7
```
Comment (agelsomini) [inline requirements.txt]: This looks out of date

**Review by agelsomini — DISMISSED** [review]
Couple of nits, but overall seems reasonable to me.  At this point this is still solely run in sim right?

**Review by agelsomini — DISMISSED** [review]

**Comment by agelsomini** [comment]
You need to resolve conflicts in requirements.txt before merging/running checks.

**Review by LandonBayer — DISMISSED** [review]

**Pull request merged**